pkg/k8s/watchers/crd.go

```diff
@@ -0,0 +1,299 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watchers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	k8sversion "github.com/cilium/cilium/pkg/k8s/version"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/option"
+
+	apiextclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (k *K8sWatcher) crdsInit(client apiextclientset.Interface) {
+	crds := newCRDState(
+		map[string]bool{
+			cnpCRD:  false,
+			ccnpCRD: false,
+			cepCRD:  false,
+			cnCRD:   false,
+			cidCRD:  false,
+			clrpCRD: false,
+		},
+	)
+
+	_, crdController := informer.NewInformer(
+		newListWatchFromClient(
+			newCRDGetter(client),
+			fields.Everything(),
+		),
+		&metav1.PartialObjectMetadata{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if pom, ok := obj.(*metav1.PartialObjectMetadata); ok {
+					crds.Lock()
+					crds.m["crd:"+pom.GetName()] = true
+					crds.Unlock()
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if pom, ok := obj.(*metav1.PartialObjectMetadata); ok {
```

Comment:
Use a helper function to translate from `obj` to `*metav1.PartialObjectMetadata`. The `obj` might be a `cache.DeletedFinalStateUnknown`. See the other functions in `pkg/k8s/factory_functions.go`